loopchain/peer/consensus_siever.py

```diff
@@ -27,88 +28,101 @@ class ConsensusSiever(ConsensusBase):
     def __init__(self, block_manager):
         super().__init__(block_manager)
         self.__block_generation_timer = None
+        self.__lock = threading.Lock()
 
     def start_timer(self, timer_service):
         self.__block_generation_timer = SlotTimer(
             TimerService.TIMER_KEY_BLOCK_GENERATE,
             conf.INTERVAL_BLOCKGENERATION,
             timer_service,
-            self.consensus
+            self.consensus,
+            self.__lock
         )
 
     def stop(self):
         self.__block_generation_timer.stop()
         self.__stop_broadcast_send_unconfirmed_block_timer()
 
     async def consensus(self):
-        block_builder = self._makeup_block()
-
-        if len(block_builder.transactions) == 0 and not conf.ALLOW_MAKE_EMPTY_BLOCK:
-            util.logger.spam(f"tx count in block({len(block_builder.transactions)})")
-            return self.__block_generation_timer.call()
-
-        peer_manager = ObjectManager().channel_service.peer_manager
+        util.logger.debug(f"-------------------consensus "
+                          f"candidate_blocks({len(self._blockmanager.candidate_blocks.blocks)})")
+        with self.__lock:
+            block_builder = self._makeup_block()
+            vote_result = None
+
+            if len(block_builder.transactions) > 0:
+                # util.logger.debug(f"-------------------consensus logic-1")
+                next_leader = ExternalAddress.fromhex(ChannelProperty().peer_id)
+
+                if self._blockchain.last_unconfirmed_block:
+                    if (len(self._blockchain.last_unconfirmed_block.body.transactions) > 0) or (
+                            len(self._blockchain.last_unconfirmed_block.body.transactions) == 0 and
+                            self._blockchain.last_unconfirmed_block.header.peer_id.hex_hx() != ChannelProperty().peer_id):
+                        # util.logger.debug(f"-------------------consensus logic-2")
+                        vote = self._blockmanager.candidate_blocks.get_vote(self._blockchain.last_unconfirmed_block.header.hash)
+                        vote_result = vote.get_result(self._blockchain.last_unconfirmed_block.header.hash.hex(), conf.VOTING_RATIO)
+                        if not vote_result:
+                            return self.__block_generation_timer.call()
+
+                        self._blockmanager.add_block(self._blockchain.last_unconfirmed_block, vote)
```

Comment:
Next unconfirmed block is not created and voted.
But unconfirmed block is added.